### PR TITLE
refactor: redpacket history

### PR DESCRIPTION
**Pull request by zhouhanseng**
- [x] deploy test-net contract
- [x] update front-end
- [x] review approved
- [x] test pass
- [x] deploy main-net contract
- [x] test pass
- [ ] merge

### CODE REVIEW

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:80]: React.CSSProperties

File: packages/maskbook/src/_locales/en/messages.json
Comment (septs) [inline packages/maskbook/src/_locales/en/messages.json:335]:
```suggestion
    "plugin_red_packet_history_duration": "Time: {{startTime}} ~ {{endTime}}",
```
we user not all in **UTC+8** time region

plz use user local time.

**Comment by septs** [comment]
After this PR merged, do i need to stop the older service?

```
https://service.r2d2.to/red-packet-history
```

CC @guanbinrui

**Comment by Tedko** [comment]
> After this PR merged, do i need to stop the older service?
> 
> ```
> https://service.r2d2.to/red-packet-history
> ```
> 
> CC @guanbinrui

keep it running for a while just in case

File: packages/maskbook/src/plugins/RedPacket/database.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/database.ts:16]: is it possible to have multiple records with the same rpid?

File: packages/maskbook/src/plugins/RedPacket/hooks/useAllRedPackets.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/hooks/useAllRedPackets.ts:6]: are you sure to design like this? it will be very expensive to transfer if the data amount is too large.

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:5]: `color`

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:6]: `backgroundColor`

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:18]: does it fit dark theme?

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:14]: does it fit dark theme?

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:44]: 🤔

File: packages/maskbook/src/plugins/RedPacket/services.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/services.ts:81]: no paging? 🤔

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:55]: then why use `theme.spacing`? Another thing, why you'd want to use `rem`? I'm curious how it work with shadow root

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:145]: IT'S NOT WORKING. `key` is a special prop name and will be removed by React.

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:249]: Why

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:258]: Why

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:18]: It fits well.

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:14]: yes

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:6]: Warning: React does not recognize the `backgroundColor` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `backgroundcolor` instead. If you accidentally passed it from a parent component, remove it from the DOM element.

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:5]: Warning: Failed prop type: Invalid prop `color` of value `rgba(44, 164, 239)` supplied to `ForwardRef(LinearProgress)`, expected one of ["primary","secondary"].

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:44]: Type '"static !important"' is not assignable to type 'Position | PropsFunc<{}, Position | undefined> | undefined'.

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:55]: 1rem = 16px in shadow root as well. But anyway let me change this to '0 16px'.

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:145]: you're right, just wrap a `<div key={...} />` for `<RedPacketInHistoryList>` solves `react-jsx-dev-runtime.development.js?9b1f:115 Warning: Each child in a list should have a unique "key" prop.
`

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:249]: I use `<strong>` tag in i18n, `"plugin_red_packet_history_claimed": "Claimed: <strong>{{claimedShares}}/{{shares}}</strong> Share",`, so if do not use dangerouslySetInnerHTML, `<strong>` would be rendered as normal text.

File: packages/maskbook/src/plugins/RedPacket/database.ts
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/database.ts:16]: No, each record has its only rpid.

File: packages/maskbook/src/plugins/RedPacket/hooks/useAllRedPackets.ts
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/hooks/useAllRedPackets.ts:6]: Currently I think it is okay, ITO has the same design since there's no user send too many red packets or ITOs. We can check the data amount for each user on subgraph. So this will be redesigned later.

File: packages/maskbook/src/plugins/RedPacket/services.ts
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/services.ts:81]: It will be redesigned later.

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:44]: hmm but why u need !important

File: packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/UI/RedPacketInList.tsx:248]: https://react.i18next.com/latest/trans-component

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (Jack-Works) [reply packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:5]: Well let me change this

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:18]: is it dark mode ready?

File: packages/maskbook/src/plugins/RedPacket/database.ts
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/database.ts:16]: so why don't you get record by `RedPacketDatabase.get`?

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Review by Jack-Works — DISMISSED** [review]
lgtm on code style

**Comment by Jack-Works** [comment]
oh wait

**Comment by Jack-Works** [comment]
1. here is confliction
2. still have some old review not resolved

File: packages/maskbook/src/plugins/RedPacket/database.ts
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/database.ts:16]: the last question @zhouhanseng

File: packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx
Comment (guanbinrui) [reply packages/maskbook/src/plugins/ITO/UI/StyledLinearProgress.tsx:18]: '#fff' => `theme.palette.common.white`

File: packages/maskbook/src/plugins/RedPacket/services.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/services.ts:85]: Check `history.chain_id === chainId` first if it's true the finding logic can be ommited.

File: packages/maskbook/src/plugins/RedPacket/database.ts
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/database.ts:16]: if you can query via `.get` plz use it instead of iterate over the database. it's O1 (hash table)

**Comment by Tedko** [comment]
How's it? @zhouhanseng @septs @Jack-Works

File: packages/maskbook/src/plugins/RedPacket/database.ts
Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/database.ts:16]: cc @zhouhanseng

**Comment by zhouhanseng** [comment]
@Tedko work on this today.

File: packages/maskbook/src/plugins/RedPacket/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/RedPacket/types.ts:42]: is it a breaking change?

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/types.ts:42]: yes, already handled it: https://github.com/DimensionDev/Maskbook/blob/8accb67a36f00124011538805797cdbf0aa6651e/packages/maskbook/src/plugins/RedPacket/hooks/useAvailabilityComputed.ts#L32

Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/types.ts:42]: I mean how do you handle the old name `ifclaimed`

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/types.ts:42]: `const isClaimed = availability.ifclaimed` => `const isClaimed = availability.claimed_amount !== '0' `

Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/types.ts:42]: So IIRC the field "ifclaimed" is actually used in the payload. did it changed too?

Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/types.ts:42]: `ifclaimed` does not exist in payload

File: packages/maskbook/src/plugins/RedPacket/services.ts
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/services.ts:85]: It can't be ommited since we need password.

File: packages/maskbook/src/plugins/RedPacket/schema.json
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/RedPacket/schema.json:13]: It is not breaking change, since `{ "additionalProperties": true }` and `"contract_version"` is removed from `"required"` field, old red-packet payload still works.

File: packages/maskbook/src/plugins/RedPacket/services.ts
Comment (zhouhanseng) [inline packages/maskbook/src/plugins/RedPacket/services.ts:53]: simplify this logic

File: packages/maskbook/src/plugins/RedPacket/contracts/useRedPacketContract.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/contracts/useRedPacketContract.ts:11]: Why V1 still available?

Comment (Jack-Works) [reply packages/maskbook/src/plugins/RedPacket/contracts/useRedPacketContract.ts:11]: why not? if so old payload will break

**Review by Jack-Works — APPROVED** [review]
currently lgtm

**Comment by Jack-Works** [comment]
merge with master. might have some type errors in it

**Comment by zhouhanseng** [comment]
@Jack-Works Notice don't merge it before mainnet contract deployed.

**Review by Jack-Works — DISMISSED** [review]
lgtm on style

**Comment by Jack-Works** [comment]
hello there are a lot of conflicts and I cannot rebase them. please merge/rebase on master @zhouhanseng

**Comment by zhouhanseng** [comment]
rebased

**Comment by septs** [comment]
@zhouhanseng plz solve all warning.

**Comment by zhouhanseng** [comment]
no warning.

**Review by Jack-Works — APPROVED** [review]
LGTM on code style

**Comment by zhouhanseng** [comment]
solve smol conflict

image

**Review by Jack-Works — APPROVED** [review]

**Comment by Tedko** [comment]
👍🏻

File: packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts:80]: Remove it.

File: packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts:82]: Remove it.

File: packages/web3-shared/src/pipes/index.ts
Comment (guanbinrui) [inline packages/web3-shared/src/pipes/index.ts:44]: Remove it?

File: packages/maskbook/src/plugins/RedPacket/hooks/useRefundCallback.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/hooks/useRefundCallback.ts:40]:
```suggestion
                .catch((error) => {
```

File: packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts:136]:
```suggestion
                .catch((error) => {
```

File: packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts:85]: Please do it in the `<TransactionDialog />` then everyone can share this logic.

File: packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts
Comment (guanbinrui) [inline packages/maskbook/src/plugins/RedPacket/hooks/useCreateCallback.ts:103]:
```suggestion
        signedPassword = signedPassword.slice(2)
```

File: packages/web3-shared/src/pipes/index.ts
Comment (zhouhanseng) [reply packages/web3-shared/src/pipes/index.ts:44]: Need it, now getChainName or resolveChainName returns `'Ethereum Mainnet'` this kind of long name, which is not compatible of redpacket payload schema.

File: packages/maskbook/src/plugins/RedPacket/hooks/useRefundCallback.ts
Comment (zhouhanseng) [reply packages/maskbook/src/plugins/RedPacket/hooks/useRefundCallback.ts:40]: terminal complaint: `TS7006: Parameter 'error' implicitly has an 'any' type.`

**Comment by AinuX** [comment]
Test passed with ETH MainNet.

**Review by nuanyang233 — APPROVED** [review]

File: packages/maskbook/src/plugins/Ethereum/UI/TransactionDialog.tsx
Comment (septs) [inline packages/maskbook/src/plugins/Ethereum/UI/TransactionDialog.tsx:149]: plz extract this code snippet as tool function

**Review by septs — APPROVED** [review]

**Pull request merged**
